compiler/rustc_hir_analysis/src/astconv/generics.rs

```diff
@@ -590,8 +591,72 @@ pub(crate) fn check_generic_arg_count(
 }
 
 /// Emits an error regarding forbidden type binding associations
-pub fn prohibit_assoc_ty_binding(tcx: TyCtxt<'_>, span: Span) {
-    tcx.sess.emit_err(AssocTypeBindingNotAllowed { span });
+pub fn prohibit_assoc_ty_binding(
+    tcx: TyCtxt<'_>,
+    span: Span,
+    segment: Option<(&hir::PathSegment<'_>, Span)>,
+) {
+    let mut err = tcx.sess.create_err(AssocTypeBindingNotAllowed { span });
+    if let Some((segment, span)) = segment {
+        if segment.args().parenthesized {
+            err.span_note(
+                span,
+                format!("this type expands to `{}`", fn_trait_to_string(tcx, segment, false)),
+            );
+        }
+    }
+    err.emit();
+}
+
+pub(crate) fn fn_trait_to_string(
```

Comment:
This function is purely for diagnostics, so it's better to move it to `compiler/rustc_hir_analysis/src/astconv/errors.rs`.